rustls/src/client/hs.rs

```diff
@@ -419,6 +376,56 @@ fn emit_client_hello_for_retry(
     }
 }
 
+fn prepare_resumption<'a>(
+    resuming: &'a Option<persist::Retrieved<ClientSessionValue>>,
+    exts: &mut Vec<ClientExtension>,
+    doing_retry: bool,
+    suite: Option<SupportedCipherSuite>,
+    cx: &mut ClientContext<'_>,
+    config: &ClientConfig,
+) -> Option<persist::Retrieved<&'a persist::Tls13ClientSessionValue>> {
+    if !config.enable_tickets {
+        return None;
+    }
+
+    // Check whether we're resuming with a non-empty ticket.
+    let resuming = match resuming {
+        Some(resuming) if !resuming.ticket().is_empty() => resuming,
+        _ => {
+            // If we don't have a ticket, request one.
+            exts.push(ClientExtension::SessionTicket(ClientSessionTicket::Request));
+            return None;
+        }
+    };
+
+    let tls13 = match resuming.map(|csv| csv.tls13()) {
+        Some(tls13) if config.supports_version(ProtocolVersion::TLSv1_3) => tls13,
+        None if config.supports_version(ProtocolVersion::TLSv1_2) => {
+            exts.push(ClientExtension::SessionTicket(ClientSessionTicket::Offer(
+                Payload::new(resuming.ticket()),
+            )));
+            return None;
+        }
+        _ => return None,
+    };
+
+    // If the server selected TLS 1.2, we can't resume.
+    let suite = match suite {
+        Some(SupportedCipherSuite::Tls13(suite)) => Some(suite),
+        #[cfg(feature = "tls12")]
+        Some(SupportedCipherSuite::Tls12(_)) => return None,
+        None => None,
+    };
+
+    // If the selected cipher suite can't select from the session's, we can't resume.
+    if let Some(suite) = suite {
+        suite.can_resume_from(tls13.suite())?;
```

Comment:
It's a bit confusing to call a non-side-effecty function just to take advantage of the `?` (and without using the Ok / Some return value). How about:

```suggestion
        if suite.can_resume_from(tls13.suite()).is_none() {
            return None
        }
```

Reply:
I don't disagree, and this is what I had at first. However, two downsides:

* Our rustfmt settings don't allow your suggested formatting, which turns it into this:

```rust
    if let Some(suite) = suite {
        if suite
            .can_resume_from(tls13.suite())
            .is_none()
        {
            return None;
        }
    }
```

* clippy warns against this approach:

```rust
warning: this block may be rewritten with the `?` operator
   --> rustls/src/client/hs.rs:422:9
    |
422 | /         if suite
423 | |             .can_resume_from(tls13.suite())
424 | |             .is_none()
425 | |         {
426 | |             return None;
427 | |         }
    | |_________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#question_mark
    = note: `#[warn(clippy::question_mark)]` on by default
help: replace it with
    |
422 ~         suite
423 +             .can_resume_from(tls13.suite())?;
    |
```

Either of these in isolation might have been fine, but given both of them I'd prefer to use the form I have here.

Reply:
Ah yeah, both good reasons to do it this way, though I disagree with clippy in this instance.